govcd/vapp.go

```diff
@@ -786,6 +798,328 @@ func (vapp *VApp) AddIsolatedNetwork(newIsolatedNetworkSettings *VappNetworkSett
 
 }
 
+// AddNetwork creates isolated or nat routed(connected to Org VDC network) network for vApp.
+// Returns pointer to types.NetworkConfigSection or error
+// If orgNetwork is nil, then isolated network created.
+func (vapp *VApp) AddNetwork(newNetworkSettings *VappNetworkSettings, orgNetwork *types.OrgVDCNetwork) (*types.NetworkConfigSection, error) {
+	task, err := vapp.AddNetworkAsync(newNetworkSettings, orgNetwork)
+	if err != nil {
+		return nil, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	vAppNetworkConfig, err := vapp.GetNetworkConfig()
+	if err != nil {
+		return nil, fmt.Errorf("error getting vApp networks: %#v", err)
+	}
+
+	return vAppNetworkConfig, nil
+}
+
+// AddNetworkAsync creates asyncronously isolated or nat routed network for vApp. Returns Task or error
+// If orgNetwork is nil, then isolated network created.
+func (vapp *VApp) AddNetworkAsync(newNetworkSettings *VappNetworkSettings, orgNetwork *types.OrgVDCNetwork) (Task, error) {
+
+	err := validateNetworkConfigSettings(newNetworkSettings)
+	if err != nil {
+		return Task{}, err
+	}
+
+	// for case when range is one ip address
+	if newNetworkSettings.DhcpSettings != nil && newNetworkSettings.DhcpSettings.IPRange != nil && newNetworkSettings.DhcpSettings.IPRange.EndAddress == "" {
+		newNetworkSettings.DhcpSettings.IPRange.EndAddress = newNetworkSettings.DhcpSettings.IPRange.StartAddress
+	}
+
+	// explicitly check if to add data, to not send any values
+	var networkFeatures *types.NetworkFeatures
+	if newNetworkSettings.DhcpSettings != nil {
+		networkFeatures = &types.NetworkFeatures{DhcpService: &types.DhcpService{
+			IsEnabled:        newNetworkSettings.DhcpSettings.IsEnabled,
+			DefaultLeaseTime: newNetworkSettings.DhcpSettings.DefaultLeaseTime,
+			MaxLeaseTime:     newNetworkSettings.DhcpSettings.MaxLeaseTime,
+			IPRange:          newNetworkSettings.DhcpSettings.IPRange},
+		}
+	}
+
+	if newNetworkSettings.FirewallEnabled != nil {
+		if networkFeatures == nil {
+			networkFeatures = &types.NetworkFeatures{}
+		}
+		networkFeatures.FirewallService = &types.FirewallService{IsEnabled: *newNetworkSettings.FirewallEnabled}
+	}
+	if newNetworkSettings.FirewallEnabled != nil && newNetworkSettings.NatEnabled != nil {
```

Comment:
Does it imply that firewall must be enabled so that NAT rules can be enabled? This looks to be handled a bit differently in `AddOrgNetworkAsync`

Reply:
made to be the same in AddOrgNetworkAsync

Reply:
After a discussion we proved that NAT cannot be enabled without firewall being enabled.